### ISSUE TITLE
Questions about fine-tuning this model on our custom dataset

### ISSUE THREAD

**Issue by booker-max** [yuliangguo/depth_any_camera#7]
Thank you for your fantastic work!
This paper demonstrates that using ERP as an intermediate representation enables models trained on perspective images to generalize well to fisheye images. 
However, I have a few questions regarding fine-tuning this model on my own data. If I already have the original fisheye images and their corresponding depth maps, can I directly fine-tune this model on the fisheye data and depth maps, or do I still need to convert them into the ERP representation in fine-tuning?

Additionally, if possible, could you provide a script for fine-tuning on ScanNetPP? 
Since this type of KB model is applicable to most fisheye cameras, having such a script would further enhance the model's performance on custom fisheye datasets.

**Comment by yuliangguo** [comment]
Thank you for this good question. Actually this also belongs to our work's
ultimate form: mixing all different types of camera types in the
training to build a foundation model. However, it has not been done yet,
due to resource limitation we need to sort out first.

For your question of fine-tuning on scannet++, yes, you need to convert the
fisheye to ERP so that the pretrained model and the new data can
communicate in the same space, under the same scale. *You can directly use
scannetpp_erp.pp as the training data loader.* However, because scannet++
itself is so huge already, if you just care about its in-domain performance
on its own test, training from scratch might be more straight-forward.

For the fine-tune script, I think it would only make sense when the
training set is small, e.g., a small portion of scannet++ to demonstrate
the fine-tune process. However, I may not guarantee it coming soon.

**Comment by booker-max** [comment]
Thank you! I look forward to mixing different types of cameras in the training process to build a foundational model. 
I like this foundational model.
I am also experimenting with training on mixed datasets, and I will fine-tune my model on the dataset based on your suggestions. Once I have the results, I will share them with you as soon as possible. Thank you!

**Comment by booker-max** [comment]
![Image](https://github.com/user-attachments/assets/4426925f-dc33-4522-b690-787fb9721a2a)

![Image](https://github.com/user-attachments/assets/192a475f-7e21-4ac8-8a88-14f8c25b5326)

These images show the conversion to ERP, the fisheye image obtained by remapping the ERP image, and my original fisheye image. I'm not sure if there is an issue with the conversion to ERP or the remapping process, because the remapped fisheye image does not match my original fisheye image."

**Comment by yuliangguo** [comment]
You need to produce your own lookup table for your fisheye parameters. Have
you done that?

**Comment by booker-max** [comment]
Yes, I recreated my own lookup table because my fisheye imaging model is the same as the one used in the ScanNet++ dataset, both following the KB model. I also processed the data using an approach similar to the ScanNet++ code. However, the remapped image I generated still does not match the original image. The specific code is as follows.

`import numpy as np
import torch
import cv2
import os
from tqdm import tqdm
import matplotlib.pyplot as plt

def generate_fisheye_grid(H, W, cam_in, distortion, output_path):
    fx, fy, cx, cy = cam_in
    k1, k2, k3, k4, *_ = distortion

    H = int(H / 2)
    W = int(W / 2)

    fx = fx / 2
    fy = fy / 2
    cx = cx / 2
    cy = cy / 2

    u, v = np.meshgrid(np.arange(W), np.arange(H))
    u = u.reshape(-1).astype(np.float32) + 0.5  # 加半像素
    v = v.reshape(-1).astype(np.float32) + 0.5
    pixels = np.stack((u, v, np.ones_like(u)), axis=0)  # [3, H*W]
    grid = torch.from_numpy(pixels)

    map_dist = []
    for ro in np.linspace(0.0, 15, 500000):
        theta = np.arctan(ro)
        theta_d = theta * (1 + k1 * theta**2 + k2 * theta**4 + k3 * theta**6 + k4 * theta**8)
        map_dist.append([ro, theta_d])
    map_dist = np.array(map_dist).astype(np.float32)
    map_dist = torch.from_numpy(map_dist).cuda()

    def chunk(grid):
        x = (grid[0, :] - cx) / fx
        y = (grid[1, :] - cy) / fy
        dist = torch.sqrt(x * x + y * y)
        indx = torch.abs(map_dist[:, 1:] - dist[None, :]).argmin(dim=0)
        x = x * map_dist[indx, 0] / map_dist[indx, 1]
        y = y * map_dist[indx, 0] / map_dist[indx, 1]
        z = 1 / torch.sqrt(1 + x**2 + y**2)
        xy = torch.stack((x, y))
        xy *= z
        return xy

    xys = []
    for i in tqdm(range(H)):
        xy = chunk(grid[:, i * W:(i + 1) * W].cuda())
        xys.append(xy.permute(1, 0))
    xys = torch.cat(xys, dim=0)

    z = torch.sqrt(1. - torch.norm(xys, dim=1, p=2) ** 2)
    isnan = z.isnan()
    z[isnan] = 1.
    pcd = torch.cat((xys, z[:, None], isnan[:, None]), dim=1)

    # 保存 npy 文件
    pcd_np = pcd.detach().cpu().numpy().reshape(H, W, 4)
    np.save(output_path, pcd_np)
    print(f"Saved fisheye grid to {output_path}")

    def cam2image(pcd):
        x = pcd[:, 0] / pcd[:, 2]
        y = pcd[:, 1] / pcd[:, 2]
        r = torch.sqrt(x**2 + y**2)
        theta = torch.arctan(r)
        theta_d = theta * (1 + k1 * theta**2 + k2 * theta**4 + k3 * theta**6 + k4 * theta**8)
        x = theta_d * x / (r + 1e-9)
        y = theta_d * y / (r + 1e-9)
        x = fx * x + cx
        y = fy * y + cy
        return x, y

    x, y = cam2image(pcd[:, :3])
    error = (x - grid[0].cuda()) ** 2 + (y - grid[1].cuda()) ** 2
    error_map = error.reshape(H, W).detach().cpu().numpy()
    error_map = np.clip(error_map, 0, 30)
    plt.imshow(error_map)
    plt.colorbar()
    plt.savefig(output_path.replace('.npy', '_error_map.png'))
    print(f"Saved error map to {output_path.replace('.npy', '_error_map.png')}")

cam_in = [
    437.2405090332031,
    437.0775146484375,
    960.3468627929688,
    721.3284301757812
]

distortion = [
    0.22122569382190704,
    -0.058465998619794846,
    0.007928100414574146,
    -0.0008349000127054751,
    0.0,
    0.0
]

rgb_image_path = ""
output_path = ""


H, W = cv2.imread(rgb_image_path).shape[:2]  
generate_fisheye_grid(H, W, cam_in, distortion, output_path)`, 

![Image](https://github.com/user-attachments/assets/c4d7fb99-57fb-429e-b051-e925da2022ee)

**Comment by yuliangguo** [comment]
OK, the cut-out region when converting to ERP looks to be out of hFOV 180 degree. It worth to note that for hFOV > 180, current ERP converion still have some limitation, some additional conversion like MEI model might be needed. However, when designing a modified ERP space,  the model need to be re-trained either.

On the other hand, I have made another update of ERP geometry making ERP converion more robust to large FOV fisheye mode. You may take a try first.

**Comment by booker-max** [comment]
Thank you for your reply. You’re right—my dataset’s HFOV does exceed 180 degrees. When I run the code to generate the fisheye_grid, the error map appears as shown in the figure. The area with the largest error is the portion beyond 180 degrees, and unfortunately, I don’t have a better solution for that at the moment. Thank you again.
![Image](https://github.com/user-attachments/assets/3a03b4b6-b5b8-41bf-b342-f474c95c5e5c)，
![Image](https://github.com/user-attachments/assets/909101b5-126e-4075-800e-7a277b8f6caa)

**Comment by yuliangguo** [comment]
Note that such lookup table is only for converting visualization to the
original camera space. The depth estimated within the erp range and the
generated point clouds are not affected.
